radixdlt/src/main/java/com/radixdlt/consensus/liveness/WeightedRotatingLeaders.java

```diff
@@ -0,0 +1,142 @@
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied.  See the License for the specific
+ * language governing permissions and limitations under the License.
+ */
+
+package com.radixdlt.consensus.liveness;
+
+import com.radixdlt.consensus.View;
+import com.radixdlt.consensus.validators.Validator;
+import com.radixdlt.consensus.validators.ValidatorSet;
+import com.radixdlt.crypto.ECPublicKey;
+import com.radixdlt.utils.UInt128;
+import com.radixdlt.utils.UInt256;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Map.Entry;
+
+/**
+ * Rotates leaders with those having more power being proposed more often
+ * in proportion to the amount of power they have.
+ *
+ * Calculation of the next leader is dependent on the weight state of the
+ * previous view and thus computing the leader for an arbitrary view can
+ * be quite expensive.
+ *
+ * We resolve this by keeping a cache of some given size of the previous
+ * views closest to the highest view calculated.
+ *
+ * This class stateful and is NOT thread-safe.
+ */
+public final class WeightedRotatingLeaders implements ProposerElection {
+
+	private final ValidatorSet validatorSet;
+	private final Comparator<Entry<Validator, UInt256>> weightsComparator;
+	private final CachingNextLeaderComputer nextLeaderComputer;
+
+	public WeightedRotatingLeaders(ValidatorSet validatorSet, Comparator<Validator> comparator, int cacheSize) {
+		this.validatorSet = validatorSet;
+		this.weightsComparator = Comparator
+			.comparing(Entry<Validator, UInt256>::getValue)
+			.thenComparing(Entry::getKey, comparator);
+		this.nextLeaderComputer = new CachingNextLeaderComputer(validatorSet, weightsComparator, cacheSize);
+	}
+
+	private static class CachingNextLeaderComputer {
+		private final ValidatorSet validatorSet;
+		private final Comparator<Entry<Validator, UInt256>> weightsComparator;
+		private final Map<Validator, UInt256> weights;
+		private final Validator[] cache;
+		private View curView;
+
+		private CachingNextLeaderComputer(ValidatorSet validatorSet, Comparator<Entry<Validator, UInt256>> weightsComparator, int cacheSize) {
+			this.validatorSet = validatorSet;
+			this.weightsComparator = weightsComparator;
+			this.weights = new HashMap<>();
+			this.cache = new Validator[cacheSize];
+			this.resetToView(View.of(0));
+		}
+
+		private Validator computeHeaviest() {
+			final Entry<Validator, UInt256> max = weights.entrySet().stream()
+				.max(weightsComparator)
+				.orElseThrow(() -> new IllegalStateException("Weights cannot be empty"));
+			return max.getKey();
+		}
+
+		private void computeNext() {
+			// Reset last leader
+			final int curIndex = (int) (this.curView.number() % cache.length);
+			final Validator curLeader = cache[curIndex];
+			weights.merge(curLeader, validatorSet.getTotalPower(), UInt256::subtract);
```

Comment:
Took me a bit to get why this works. Please add a few comments to the lines where weights are modified to explain the idea (reduce by total and then accumulate with power so that the leader will rise back to the top with a pace corresponding to their relative power)

Reply:
good point